### PR TITLE
Refactor RetrieveValidatedConfigInfo to reduce its complexity

### PR DESCRIPTION
**Pull request by ping035627**
What type of PR is this?
/kind cleanup

What this PR does / why we need it:
Cyclomatic complexity 21 of function RetrieveValidatedConfigInfo(), it's so high, I refactor it to reduce its complexity. I extract two function from "RetrieveValidatedConfigInfo", “getInsecureKubeconfigBytes“ and “getSecureKubeconfig”.

Which issue(s) this PR fixes:

Fixes #

Special notes for your reviewer:

Does this PR introduce a user-facing change?:

/release-note-none
/priority backlog

### CODE REVIEW

**Comment by ping035627** [comment]
@kubernetes/sig-cluster-lifecycle-pr-reviews

**Review by neolit123 — COMMENTED** [review]
/ok-to-test
/test pull-kubernetes-e2e-kind

**Comment by neolit123** [comment]
@ping035627 
> Cyclomatic complexity 21 of function RetrieveValidatedConfigInfo(), it's so high

out of interest, what software are you using to determine that?

**Comment by ping035627** [comment]
@neolit123 gometalinter(gocyclo), our project has some requirements for static code checking, such as cyclomatic complexity less 12, eliminate HIGH and MEDIUM warnings of gosec, etc.

**Comment by neolit123** [comment]
> @neolit123 gometalinter(gocyclo), our project has some requirements for static code checking, such as cyclomatic complexity less 12, eliminate HIGH and MEDIUM warnings of gosec, etc.

thank you for the details.

File: cmd/kubeadm/app/discovery/token/token.go [hunk 1]
```diff
@@ -171,6 +123,74 @@ func RetrieveValidatedConfigInfo(cfg *kubeadmapi.JoinConfiguration) (*clientcmda
...
+			bootstrapapi.KubeConfigKey, bootstrapapi.ConfigMapClusterInfo)
+	}
+	detachedJWSToken, ok := insecureClusterInfo.Data[bootstrapapi.JWSSignatureKeyPrefix+token.ID]
+	if !ok || len(detachedJWSToken) == 0 {
+		return nil, errors.Errorf("token id %q is invalid for this cluster or it has expired. Use \"kubeadm token create\" on the control-plane node to create a new valid token", token.ID)
+	}
+	if !bootstrap.DetachedTokenIsValid(detachedJWSToken, insecureKubeconfigString, token.ID, token.Secret) {
+		return nil, errors.New("failed to verify JWS signature of received cluster info object, can't trust this API Server")
+	}
+	insecureKubeconfigBytes := []byte(insecureKubeconfigString)
```
Comment (rosti) [inline cmd/kubeadm/app/discovery/token/token.go]: You can move this to the return below.

File: cmd/kubeadm/app/discovery/token/token.go [hunk 2]
```diff
@@ -171,6 +123,74 @@ func RetrieveValidatedConfigInfo(cfg *kubeadmapi.JoinConfiguration) (*clientcmda
...
+		insecureClusterInfo, err = insecureClient.CoreV1().ConfigMaps(metav1.NamespacePublic).Get(bootstrapapi.ConfigMapClusterInfo, metav1.GetOptions{})
+		if err != nil {
+			klog.V(1).Infof("[discovery] Failed to request cluster info, will try again: [%s]\n", err)
+			return false, nil
+		}
+		return true, nil
+	})
+
+	// Validate the MAC on the kubeconfig from the ConfigMap and load it
+	insecureKubeconfigString, ok := insecureClusterInfo.Data[bootstrapapi.KubeConfigKey]
```
Comment (rosti) [inline cmd/kubeadm/app/discovery/token/token.go]: From the start of the func (the call of `ToClientSet`) to this line (inclusive), the code looks mostly the same with the start of `getSecureKubeconfig`.
Can you extract the common part in another helper (to be used by these couple of funcs)?

**Review by rosti — COMMENTED** [review]
Nice work @ping035627 !

File: cmd/kubeadm/app/discovery/token/token.go [hunk 3]
```diff
@@ -131,32 +104,11 @@ func RetrieveValidatedConfigInfo(cfg *kubeadmapi.JoinConfiguration) (*clientcmda
...
 
-		// Pull the kubeconfig from the securely-obtained ConfigMap and validate that it's the same as what we found the first time
-		secureKubeconfigBytes := []byte(secureClusterInfo.Data[bootstrapapi.KubeConfigKey])
-		if !bytes.Equal(secureKubeconfigBytes, insecureKubeconfigBytes) {
-			return nil, errors.Errorf("the second kubeconfig from the %s configmap (using validated TLS) was different from the first", bootstrapapi.ConfigMapClusterInfo)
-		}
-
-		secureKubeconfig, err := clientcmd.Load(secureKubeconfigBytes)
+		// Now that we know the proported cluster CA, connect back a second time validating with that CA
+		secureBootstrapConfig := buildSecureBootstrapKubeConfig(endpoint, clusterCABytes, clusterName)
```
Comment (fabriziopandini) [inline cmd/kubeadm/app/discovery/token/token.go]: secureBootstrapConfig seems used only inside getSecureKubeconfig....

File: cmd/kubeadm/app/discovery/token/token.go [hunk 4]
```diff
@@ -171,6 +123,74 @@ func RetrieveValidatedConfigInfo(cfg *kubeadmapi.JoinConfiguration) (*clientcmda
 	return baseKubeConfig, nil
 }
 
+//getInsecureKubeconfigBytes is used to return insecureKubeconfigBytes from insecureBootstrapConfig and token
+func getInsecureKubeconfigBytes(insecureBootstrapConfig *clientcmdapi.Config, token *kubeadmapi.BootstrapTokenString) ([]byte, error) {
```
Comment (fabriziopandini) [inline cmd/kubeadm/app/discovery/token/token.go]: Please elaborate comment to the new functions
This refactor is a great opportunity to better document how kubeadm validates cluster identity

**Review by fabriziopandini — COMMENTED** [review]
@ping035627 thanks for this refactor. Great to see another PR from you
Apart from pending comments, IMO new functions should have better comments because I would that the work performed by the code remains clear and consistent no matter if we are splitting into more func

File: cmd/kubeadm/app/discovery/token/token.go (hunk 3 above)
Comment (ping035627) [reply cmd/kubeadm/app/discovery/token/token.go]: secureBootstrapConfig used only inside getSecureKubeconfig, but if we put it in getSecureKubeconfig, we need to pass more parameters, so I put it out of getSecureKubeconfig.

File: cmd/kubeadm/app/discovery/token/token.go (hunk 4 above)
Comment (ping035627) [reply cmd/kubeadm/app/discovery/token/token.go]: Thanks for your reminding, I will improve the comment.

File: cmd/kubeadm/app/discovery/token/token.go (hunk 2 above)
Comment (ping035627) [reply cmd/kubeadm/app/discovery/token/token.go]: Thanks for your reminding, I have modified.

File: cmd/kubeadm/app/discovery/token/token.go [hunk 5]
```diff
@@ -171,6 +119,78 @@ func RetrieveValidatedConfigInfo(cfg *kubeadmapi.JoinConfiguration) (*clientcmda
...
+		return nil, err
+	}
+
+	var clusterInfo *v1.ConfigMap
+	wait.PollImmediateInfinite(constants.DiscoveryRetryInterval, func() (bool, error) {
+		var err error
+		clusterInfo, err = client.CoreV1().ConfigMaps(metav1.NamespacePublic).Get(bootstrapapi.ConfigMapClusterInfo, metav1.GetOptions{})
+		if err != nil {
+			klog.V(1).Infof("[discovery] Failed to request cluster info, will try again: [%s]\n", err)
+			return false, nil
```
Comment (Klaven) [inline cmd/kubeadm/app/discovery/token/token.go]: did you want to return `err` here instead of nil?

**Review by Klaven — COMMENTED** [review]
Thanks for the PR!

File: cmd/kubeadm/app/discovery/token/token.go (hunk 5 above)
Comment (ping035627) [reply cmd/kubeadm/app/discovery/token/token.go]: no, it will try again.

File: cmd/kubeadm/app/discovery/token/token.go [hunk 6]
```diff
@@ -171,6 +119,78 @@ func RetrieveValidatedConfigInfo(cfg *kubeadmapi.JoinConfiguration) (*clientcmda
...
+	secureKubeconfig, err := clientcmd.Load(secureKubeconfigBytes)
+	if err != nil {
+		return nil, errors.Wrapf(err, "couldn't parse the kubeconfig file in the %s configmap", bootstrapapi.ConfigMapClusterInfo)
+	}
+	return secureKubeconfig, nil
+}
+
+// getClusterInfoFromConfig connects to the API Server and tries to fetch the cluster-info ConfigMap
+func getClusterInfoFromConfig(config *clientcmdapi.Config) (*v1.ConfigMap, error) {
+	client, err := kubeconfigutil.ToClientSet(config)
```
Comment (rosti) [inline cmd/kubeadm/app/discovery/token/token.go]: On a second thought, to make things easier to test with a fake client set, we should transform the kubeconfigs into clientsets directly into `RetrieveValidatedConfigInfo` and pass them as such to the new funcs. Hence, the calls to `ToClientSet` should be happening in `RetrieveValidatedConfigInfo`.
Sorry for forgetting this detail earlier on.

**Review by rosti — COMMENTED** [review]
Thanks for the update @ping035627 and sorry for the review delay!
I like how things are turning in the code, but we also need to test the three new funcs. You can employ fake client sets for this.

File: cmd/kubeadm/app/discovery/token/token.go (hunk 6 above)
Comment (ping035627) [reply cmd/kubeadm/app/discovery/token/token.go]: ok, I have reverted the calls to ToClientSet in RetrieveValidatedConfigInfo.

File: cmd/kubeadm/app/discovery/token/token.go [hunk 7]
```diff
@@ -86,23 +86,9 @@ func RetrieveValidatedConfigInfo(cfg *kubeadmapi.JoinConfiguration) (*clientcmda
...
-		detachedJWSToken, ok := insecureClusterInfo.Data[bootstrapapi.JWSSignatureKeyPrefix+token.ID]
-		if !ok || len(detachedJWSToken) == 0 {
-			return nil, errors.Errorf("token id %q is invalid for this cluster or it has expired. Use \"kubeadm token create\" on the control-plane node to create a new valid token", token.ID)
-		}
-		if !bootstrap.DetachedTokenIsValid(detachedJWSToken, insecureKubeconfigString, token.ID, token.Secret) {
-			return nil, errors.New("failed to verify JWS signature of received cluster info object, can't trust this API Server")
-		}
-		insecureKubeconfigBytes := []byte(insecureKubeconfigString)
-		insecureConfig, err := clientcmd.Load(insecureKubeconfigBytes)
+		insecureKubeconfigBytes, insecureConfig, err := getInsecureConfig(insecureClusterInfo, token)
```
Comment (rosti) [inline cmd/kubeadm/app/discovery/token/token.go]: The idea of my last proposition was to have a call like:
```
... := getInsecureConfig(insecureClient, token)
```

and have the `PollImmediateInfinite` stuff from above should be shared between `getSecureKubeconfig` and `getInsecureKubeconfig` (be in a third func).

Comment (ping035627) [reply cmd/kubeadm/app/discovery/token/token.go]: I see, I just think it's a little small just PollImmediateInfinite as a func. While it‘s the better to share it between getSecureKubeconfig and getInsecureKubeconfig. I've revised it. Thanks!

Comment (ping035627) [reply cmd/kubeadm/app/discovery/token/token.go]: @rosti hi, how about this pr? Is there anything else to be modified?

**Comment by ping035627** [comment]
@rosti hi, how about this pr? Is there anything else to be modified?

**Pull request closed without merging**